beacon-chain/sync/service_test.go

```diff
@@ -685,3 +696,221 @@ func TestProcessSameActiveState(t *testing.T) {
+		Peer: p2p.Peer{},
+		Data: blockResponse,
+	}
+
+	ss.blockBuf <- msg1
+
+	ss.cancel()
+	<-exitRoutine
+
+	var hash [32]byte
+	copy(hash[:], blockResponse.Block.CrystallizedStateHash)
+
+	if hash != ss.initialCrystallizedStateHash {
+		t.Fatalf("Crystallized state hash not updated: %x", blockResponse.Block.CrystallizedStateHash)
+	}
+
+	hook.Reset()
+
+}
+
+func TestSavingBlocksInSync(t *testing.T) {
+	hook := logTest.NewGlobal()
+	interval := time.Duration(100000)
+
+	cfg := Config{BlockBufferSize: 0, CrystallizedStateBufferSize: 0, SyncPollingInterval: interval}
+	ms := &mockChainService{}
+	ss := NewSyncService(context.Background(), cfg, &mockP2P{}, ms)
+
+	exitRoutine := make(chan bool)
+
+	go func() {
+		ss.initialSync(ss.ctx.Done())
+		exitRoutine <- true
+	}()
+
+	generichash := make([]byte, 32)
+	generichash[0] = 'a'
+
+	crystallizedState := &pb.CrystallizedState{
+		LastFinalizedEpoch: 99,
+	}
+
+	stateResponse := &pb.CrystallizedStateResponse{
+		CrystallizedState: crystallizedState,
+	}
+
+	incorrectState := &pb.CrystallizedState{
+		LastFinalizedEpoch: 9,
+		LastJustifiedEpoch: 20,
+	}
+
+	incorrectStateResponse := &pb.CrystallizedStateResponse{
+		CrystallizedState: incorrectState,
+	}
+
+	crystallizedStateHash, err := types.NewCrystallizedState(crystallizedState).Hash()
+	if err != nil {
+		t.Fatalf("unable to get hash of crystallized state: %v", err)
+	}
+
+	block := &pb.BeaconBlock{
+		MainChainRef:          []byte{1, 2, 3},
+		ParentHash:            generichash,
+		SlotNumber:            uint64(20),
+		CrystallizedStateHash: crystallizedStateHash[:],
+	}
+
+	blockResponse := &pb.BeaconBlockResponse{
+		Block: block,
+	}
+
+	msg1 := p2p.Message{
+		Peer: p2p.Peer{},
+		Data: blockResponse,
+	}
+
+	msg2 := p2p.Message{
+		Peer: p2p.Peer{},
+		Data: incorrectStateResponse,
+	}
+
+	ss.blockBuf <- msg1
+	ss.crystallizedStateBuf <- msg2
+
+	if ss.currentSlotNumber == incorrectStateResponse.CrystallizedState.LastFinalizedEpoch {
+		t.Fatalf("Crystallized state updated incorrectly: %x", ss.currentSlotNumber)
+	}
+
+	msg2.Data = stateResponse
+
+	ss.crystallizedStateBuf <- msg2
+
+	if crystallizedStateHash != ss.initialCrystallizedStateHash {
+		t.Fatalf("Crystallized state hash not updated: %x", blockResponse.Block.CrystallizedStateHash)
+	}
+
+	blockResponse.Block.SlotNumber = 30
+	msg1.Data = blockResponse
+	ss.blockBuf <- msg1
+
+	if stateResponse.CrystallizedState.GetLastFinalizedEpoch() != ss.currentSlotNumber {
+		t.Fatalf("slotnumber saved when it was not supposed too: %v", stateResponse.CrystallizedState.GetLastFinalizedEpoch())
+	}
+
+	blockResponse.Block.SlotNumber = 100
+	ss.blockBuf <- msg1
+
+	if blockResponse.Block.GetSlotNumber() != ss.currentSlotNumber {
+		t.Fatalf("slotnumber not updated despite receiving a valid block: %v", blockResponse.Block.GetSlotNumber())
+	}
+
+	ss.cancel()
+	<-exitRoutine
+
+	hook.Reset()
+
+}
+
+func TestTimeChan(t *testing.T) {
+	hook := logTest.NewGlobal()
+	interval := time.Duration(1000)
```

Comment:
This can be avoided by passing the channel into the goroutine. We should really avoid adding extra time to the tests like this. They add up quickly.